bigframes/dataframe.py

```diff
@@ -1123,19 +1123,18 @@ def _assign_single_item(
             # local_df is likely (but not guarunteed) to be cached locally
             # since the original list came from memory and so is probably < MAX_INLINE_DF_SIZE
 
-            this_offsets_col_id = bigframes.core.guid.generate_guid()
-            this_expr = self._get_block()._expr.promote_offsets(this_offsets_col_id)
-            block = blocks.Block(
-                expr=this_expr,
-                index_labels=self.index.names,
-                index_columns=self._block.index_columns,
-                column_labels=[this_offsets_col_id] + list(self._block.value_columns),
-            )  # offsets are temporarily the first value column, label set to id
-            this_df_with_offsets = DataFrame(data=block)
-            join_result = this_df_with_offsets.join(
-                other=local_df, on=this_offsets_col_id, how="left"
+            new_column_block = local_df._block
+            original_index_column_ids = self._block.index_columns
+            self_block = self._block.reset_index(drop=False)
+            result_index, (get_column_left, get_column_right) = self_block.index.join(
```

Comment:
We should probably add some "join_by_offset" operator at some point as we do this kind of thing in a few places.